src/psyclone/dynamo0p3.py

```diff
@@ -8831,6 +8837,132 @@ def stencil(self, value):
+        if proxy:
+            proxy_str = "_proxy"
+
+        def _find_or_create_type(mod_name, type_name):
+            '''
+            Utility to find or create a TypeSymbol with the supplied name,
+            imported from the module of the supplied name.
+
+            :param str mod_name: the name of the module from which the \
+                                 TypeSymbol should be imported.
+            :param str type_name: the name of the derived type for which to \
+                                  create a TypeSymbol.
+
+            :returns: the symbol for the requested type.
+            :rtype: :py:class:`psyclone.psyir.symbols.TypeSymbol`
+
+            '''
+            try:
+                fld_type = symbol_table.lookup(type_name)
+            except KeyError:
+                try:
+                    fld_mod_container = symbol_table.lookup(mod_name)
+                except KeyError:
+                    fld_mod_container = ContainerSymbol(mod_name)
+                    root_table.add(fld_mod_container)
+                fld_type = TypeSymbol(
+                    type_name, DeferredType(),
+                    interface=GlobalInterface(fld_mod_container))
+                root_table.add(fld_type)
+            return fld_type
+
+        if self.is_scalar:
+
+            api_config = Config.get().api_conf("dynamo0.3")
+
+            if self.intrinsic_type == 'real':
+                kind_name = api_config.default_kind["real"]
+                prim_type = ScalarType.Intrinsic.REAL
+            elif self.intrinsic_type == 'integer':
+                kind_name = api_config.default_kind["integer"]
+                prim_type = ScalarType.Intrinsic.INTEGER
+            elif self.intrinsic_type == 'logical':
+                kind_name = api_config.default_kind["logical"]
+                prim_type = ScalarType.Intrinsic.BOOLEAN
+            else:
+                raise NotImplementedError(
+                    "Unsupported scalar type '{0}'".format(
+                        self.intrinsic_type))
+            try:
+                kind_symbol = symbol_table.lookup(kind_name)
+            except KeyError:
+                try:
+                    constants_container = symbol_table.lookup(
+                        "constants_mod")
+                except KeyError:
+                    constants_container = ContainerSymbol("constants_mod")
+                    root_table.add(constants_container)
+                kind_symbol = DataSymbol(
+                        kind_name, INTEGER_SINGLE_TYPE,
+                        interface=GlobalInterface(constants_container))
+                root_table.add(kind_symbol)
+            return ScalarType(prim_type, kind_symbol)
+
+        elif self.is_field:
+
+            # Find or create the TypeSymbol for the appropriate field type.
+            if self.intrinsic_type == 'real':
+                mod_name = "field_mod"
+                type_name = "field{0}_type".format(proxy_str)
```

Comment:
I look forward to us supporting symbols properly so we don;t need to hardcode names.